src/librustc_codegen_llvm/common.rs

```diff
@@ -356,6 +369,19 @@ pub fn val_ty(v: &'ll Value) -> &'ll Type {
     }
 }
 
+pub fn val_addr_space_opt(v: &'ll Value) -> Option<AddrSpaceIdx> {
+    let ty = val_ty(v);
+    if ty.kind() == TypeKind::Pointer {
+        Some(ty.address_space())
+    } else {
+        None
+    }
+}
+
+pub fn val_addr_space(v: &'ll Value) -> AddrSpaceIdx {
+    val_addr_space_opt(v).unwrap_or_default()
```

Comment:
Reflecting LLVM values should be avoided as much as possible. New uses of `val_ty` shouldn't be added.